### PR TITLE
[1.21] Add more guards to fake players

### LINKED ISSUE #1320: Fake Player are able to mount entities, afterwards causing them to despawn when server stops.
**Minecraft Version:** 1.20.1 and up (and lower mayhaps)

**NeoForge Version:** 47.1.106

**Logs:** https://mclo.gs/f9BSXNw not needed, but here anyway.

**Steps to Reproduce:**
 1. Setup a fake player with the right click action (In this example it is provided by the create mod deployer)
 2. Spawn a ride-able entity it is right clicking and let it mount the said entity
 3. Disconnect from the world
 4. Join back and observe the rid-able entity no longer existing

**Description of issue:**
The Fake Player mounts the entity, which by itself shouldn't be a thing, as it functions in pretty unintended ways, but it also does mean that when the server stops, and the Fake Player "disconnects", it casues the entity to go with it, and because the fake player never "reconnects", it causes the mount to be basically deleted.

I managed to solve this issue by using the "EntityMountEvent", where I cancelled it when a FakePlayer is passed into it.

```java
    @SubscribeEvent
    public static void onEntityMount(EntityMountEvent event) {
        if (event.getEntity() instanceof FakePlayer) {
            event.setCanceled(true);
        }
    }
```

### PR DESCRIPTION
**Pull request by XFactHD**
This PR adds more guards to fake players, specifically by no-oping menu and horse inventory opening and preventing fake players from riding other entities.
Also adds a method to check whether a player is a fake player without having to `instanceof FakePlayer`.

Fixes #1320

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/common/extensions/IPlayerExtension.java [hunk 1]
```diff
@@ -81,4 +81,11 @@ default boolean mayFly() {
         // TODO 1.20.5: consider forcing mods to use the attribute
         return self().getAbilities().mayfly || self().getAttributeValue(NeoForgeMod.CREATIVE_FLIGHT) > 0;
     }
+
+    /**
+     * {@return whether this player is a fake player}
```
Comment (TelepathicGrunt) [inline src/main/java/net/neoforged/neoforge/common/extensions/IPlayerExtension.java]: Should this javadoc contain a link to FakePlayer class so people know what it means by fake player

**Review by TelepathicGrunt — APPROVED** [review]

File: src/main/java/net/neoforged/neoforge/common/extensions/IPlayerExtension.java [hunk 2]
```diff
@@ -81,4 +82,11 @@ default boolean mayFly() {
         // TODO 1.20.5: consider forcing mods to use the attribute
         return self().getAbilities().mayfly || self().getAttributeValue(NeoForgeMod.CREATIVE_FLIGHT) > 0;
     }
+
+    /**
+     * {@return whether this player is a fake player, such as {@link FakePlayer}}
+     */
+    default boolean isFakePlayer() {
```
Comment (KnightMiner) [inline src/main/java/net/neoforged/neoforge/common/extensions/IPlayerExtension.java:89]: Thought to consider: would it be worth putting this on `LivingEntity`? Occasionally there is an interaction where you wish to check if its a fake player but are fine with a non-player living entity. With this method, you'd still have to do something like `living instanceof Player player && player.isFakePlayer()` in that case. I cannot think of a case where "fake player or non-player entity" is the desire.

The idea could be extended to other "fake entities" though I cannot think of a usecase of such a concept. If that were a useful thing then a name like `isFakeEntity` would be better.

Comment (HenryLoenwind) [reply src/main/java/net/neoforged/neoforge/common/extensions/IPlayerExtension.java:89]: fake entities are usually real ones that just haven't been added to the world. Unlike with the player object, it's easy to do that with other entities. But yes, I can see pulling up "isFake()" to Entity making sense even if no one ever made a fake non-player entity.

**Review by marchermans — APPROVED** [review]

File: src/main/java/net/neoforged/neoforge/common/extensions/IPlayerExtension.java (hunk 2 above)
Comment (XFactHD) [reply src/main/java/net/neoforged/neoforge/common/extensions/IPlayerExtension.java:89]: I'm not really a fan of pulling something so closely tied to players up into `LivingEntity`

Comment (marchermans) [reply src/main/java/net/neoforged/neoforge/common/extensions/IPlayerExtension.java:89]: Agreed if you need to know if it is a fake player, you should check for a player first. Pattern instanceof is your friend here.

**Pull request merged**